core-backend/common/src/lib.rs

```diff
@@ -246,83 +247,99 @@ pub trait BackendExt: EnvExt + CountersOwner {
+impl<EnvSystemError: Display, PrepareMemoryError: Display>
+    EnvironmentError<EnvSystemError, PrepareMemoryError>
+{
+    pub fn from_infallible(err: EnvironmentError<EnvSystemError, Infallible>) -> Self {
         match err {
-            EnvironmentExecutionError::System(err) => Self::System(err),
-            EnvironmentExecutionError::PrepareMemory(_, err) => match err {},
-            EnvironmentExecutionError::Actor(gas_amount, s) => Self::Actor(gas_amount, s),
+            EnvironmentError::System(err) => Self::System(err),
+            EnvironmentError::PrepareMemory(_, err) => match err {},
+            EnvironmentError::Actor(gas_amount, s) => Self::Actor(gas_amount, s),
         }
     }
 }
 
-type EnvironmentBackendReport<Env, EP> =
-    BackendReport<<Env as Environment<EP>>::Memory, <Env as Environment<EP>>::Ext>;
+type EnvironmentBackendReport<Env, EntryPoint> =
+    BackendReport<<Env as Environment<EntryPoint>>::Memory, <Env as Environment<EntryPoint>>::Ext>;
 
-pub type EnvironmentExecutionResult<T, Env, EP> = Result<
-    EnvironmentBackendReport<Env, EP>,
-    EnvironmentExecutionError<<Env as Environment<EP>>::Error, T>,
+pub type EnvironmentExecutionResult<PrepareMemoryError, Env, EntryPoint> = Result<
+    EnvironmentBackendReport<Env, EntryPoint>,
+    EnvironmentError<<Env as Environment<EntryPoint>>::SystemError, PrepareMemoryError>,
 >;
 
-pub trait Environment<EP = DispatchKind>: Sized
+pub trait Environment<EntryPoint = DispatchKind>: Sized
 where
-    EP: WasmEntry,
+    EntryPoint: WasmEntryPoint,
 {
-    type Ext: BackendExt + 'static;
+    type Ext: BackendExternalities + 'static;
 
     /// Memory type for current environment.
     type Memory: Memory;
 
-    /// An error issues in environment.
-    type Error: Debug + Display;
+    /// That's an error which originally comes from the primary
+    /// wasm execution environment (set by wasmi or sandbox).
+    /// So it's not the error of the `Self` itself, it's a kind
+    /// of wrapper over the underlying executor error.
+    type SystemError: Debug + Display;
 
     /// 1) Instantiates wasm binary.
     /// 2) Creates wasm memory
-    /// 3) Runs `pre_execution_handler` to fill the memory before running instance.
+    /// 3) Runs `prepare_memory` to fill the memory before running instance.
     /// 4) Instantiate external funcs for wasm module.
     fn new(
         ext: Self::Ext,
         binary: &[u8],
-        entry_point: EP,
+        entry_point: EntryPoint,
         entries: BTreeSet<DispatchKind>,
         mem_size: WasmPage,
-    ) -> Result<Self, EnvironmentExecutionError<Self::Error, Infallible>>;
+    ) -> Result<Self, EnvironmentError<Self::SystemError, Infallible>>;
 
     /// Run instance setup starting at `entry_point` - wasm export function name.
-    fn execute<F, T>(self, pre_execution_handler: F) -> EnvironmentExecutionResult<T, Self, EP>
+    fn execute<PrepareMemory, PrepareMemoryError>(
```

Comment:
As for me, we could leave the `F` and `E` generics here as well-known and readable within the context.
```suggestion
    fn execute<F, E>(
```

Reply:
I can leave `F`, but I'd like to move from `E` as it was used for errors, environment, ext. what do you think?

Reply:
agreed to resolve